### PR TITLE
Extending modern string utils to include split and strip

### PR DESCRIPTION
**Pull request by PDoakORNL**
## Proposed changes

Adds a strip and split function that support `string-view`
Existing code does not use this but it is required for #4299 

## What type(s) of changes does this code introduce?
- New feature
- Testing changes (e.g. new unit/integration/performance tests)
- Documentation changes

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
Leconte

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes This PR is up to date with current the current state of 'develop'
- Yes Code added or changed in the PR has been clang-formatted
- Yes This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by prckent** [comment]
Test this please

File: src/Utilities/tests/test_ModernStringUtils.cpp [hunk 1]
```diff
@@ -31,4 +31,61 @@ TEST_CASE("ModernStringUtils_strToLower", "[utilities]")
         "\xab"
         "not_just_ascii");
 }
+
+TEST_CASE("ModernStringUtils_split", "[utilities]")
```
Comment (markdewing) [inline src/Utilities/tests/test_ModernStringUtils.cpp:35]: There should be some tests for edge cases - empty string, string with no delimiters, string with empty delimiter, string that is only delimiters.

File: src/Utilities/tests/test_ModernStringUtils.cpp [hunk 2]
```diff
@@ -31,4 +31,61 @@ TEST_CASE("ModernStringUtils_strToLower", "[utilities]")
...
+}
+
+TEST_CASE("ModernStringUtils_string2Real", "[utilities]")
+{
+  std::string_view svalue{"101.52326626"};
+  double value = string2Real<double>(svalue);
+  CHECK(value == Approx(101.52326626));
 } // namespace qmcplusplus
+
+TEST_CASE("ModernStringUtils_strip", "[utilities]")
```
Comment (markdewing) [inline src/Utilities/tests/test_ModernStringUtils.cpp:68]: This should also have some more edge case tests - empty string, string that is only whitespace.

Comment (PDoakORNL) [reply src/Utilities/tests/test_ModernStringUtils.cpp:89]: Added

File: src/Utilities/for_testing/NativeInitializerPrint.hpp [hunk 3]
```diff
@@ -39,4 +39,16 @@ std::ostream& operator<<(std::ostream& out, const NativePrint<std::vector<T>>& n
   return out;
 }
 
+template<class T>
+std::ostream& operator<<(std::ostream& out, const NativePrint<Vector<T>>& np_vec)
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/Utilities/for_testing/NativeInitializerPrint.hpp:65]: Added file scope documentation.

[the remark this answers is not in the document]
Comment (ye-luo) [reply src/Utilities/for_testing/NativeInitializerPrint.hpp:65]: Thanks

File: src/Utilities/ModernStringUtils.hpp [hunk 4]
```diff
@@ -30,6 +34,36 @@ namespace qmcplusplus
...
+ *  calls c++ string to real conversion based on T's precision template<typename T>
+ */
+template<typename T>
+inline T string2Real(const std::string_view svalue) {
+  static_assert(std::is_floating_point_v<T>);
+      T result;
+// full support for floating point from char not present until stdlibc++ aligned with gcc11
+// there is still not from_char for floats as of libc++ 15
+#if _GLIBCXX_RELEASE > 10
+  auto [prt, ec] = std::from_chars(svalue.data(), svalue.data() + svalue.size(), result, std::chars_format::general);
```
Comment (ye-luo) [inline src/Utilities/ModernStringUtils.hpp:56]: What is the benefit of this special case? Faster? I don't mind keep only the universal working way although a bit slow.

Comment (PDoakORNL) [reply src/Utilities/ModernStringUtils.hpp:56]: It's c++ 17 and the point of this file is to use the best practice for the current standard requirement of the code.  But here we have to make an except due to incomplete support for the std library.

The new way does not require a mutable string to be constructed and in general fits with the string_view and current c++ io framework the other is a c function.

Comment (ye-luo) [reply src/Utilities/ModernStringUtils.hpp:56]: C++17 language yes. C++17 standard library, if it is a hassle due to implementation issue, I prefer using a dumb method. Is your new way safer because of error handling? That could shift my preference.

Comment (PDoakORNL) [reply src/Utilities/ModernStringUtils.hpp:56]: Yes it is able to clearly indicate an error.  atof really can't do anything you get 0.0 or UB.  

So with the c++17 way if it can't do the conversion or its out of range for the floating point type we can throw.  This means we can deal with bad input better.

**Review by markdewing — APPROVED** [review]

**Comment by ye-luo** [comment]
Test this please
